File: testers/avro/java-async/generated-and-checked-in/com/example/events/producer/TreeNodeProducer.java
package com.example.events.producer;

import com.example.events.TreeNode;
import com.example.events.header.StandardHeaders;
import java.util.concurrent.CompletableFuture;
import org.apache.kafka.clients.producer.Producer;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.clients.producer.RecordMetadata;

/** Type-safe producer for tree-node topic */
public record TreeNodeProducer(Producer<String, TreeNode> producer, String topic)
    implements AutoCloseable {
  public TreeNodeProducer(Producer<String, TreeNode> producer) {
    this(producer, "tree-node");
  }

  public TreeNodeProducer withProducer(Producer<String, TreeNode> producer) {
    return new TreeNodeProducer(producer, topic);
  }

  public TreeNodeProducer withTopic(String topic) {
    return new TreeNodeProducer(producer, topic);
  }

  /** Close the producer */
  @Override
  public void close() {
    producer.close();
  }

  /** Send a message to the topic asynchronously */
  public CompletableFuture<RecordMetadata> send(String key, TreeNode value) {
    return CompletableFuture.supplyAsync(
            () -> {
              CompletableFuture<RecordMetadata> future = new CompletableFuture<>();
              producer.send(
                  new ProducerRecord<String, TreeNode>(topic, key, value),
                  (result, exception) -> {
                    if (exception != null) {
                      future.completeExceptionally(exception);
                    } else {
                      future.complete(result);
                    }
                  });
              return future;
            })
        .thenCompose(f -> f);
  }

  /** Send a message with headers to the topic asynchronously */
  public CompletableFuture<RecordMetadata> send(
      String key, TreeNode value, StandardHeaders headers) {
    return CompletableFuture.supplyAsync(
            () -> {
              CompletableFuture<RecordMetadata> future = new CompletableFuture<>();
              producer.send(
                  new ProducerRecord<String, TreeNode>(
                      topic, null, key, value, headers.toHeaders()),
                  (result, exception) -> {
                    if (exception != null) {
                      future.completeExceptionally(exception);
                    } else {
                      future.complete(result);
                    }
                  });
              return future;
            })
        .thenCompose(f -> f);
  }
}

File: testers/avro/kotlin/generated-and-checked-in/com/example/events/producer/TreeNodeProducer.kt
package com.example.events.producer

import com.example.events.TreeNode
import com.example.events.header.StandardHeaders
import java.io.Closeable
import java.util.concurrent.Future
import org.apache.kafka.clients.producer.Producer
import org.apache.kafka.clients.producer.ProducerRecord
import org.apache.kafka.clients.producer.RecordMetadata

/** Type-safe producer for tree-node topic */
data class TreeNodeProducer(
  val producer: Producer<kotlin.String, TreeNode>,
  val topic: kotlin.String = "tree-node"
) : Closeable {
  /** Close the producer */
  override fun close() {
    producer.close()
  }

  /** Send a message to the topic */
  fun send(
    key: kotlin.String,
    value: TreeNode
  ): Future<RecordMetadata> {
    return producer.send(ProducerRecord<kotlin.String, TreeNode>(topic, key, value))
  }

  /** Send a message with headers to the topic */
  fun send(
    key: kotlin.String,
    value: TreeNode,
    headers: StandardHeaders
  ): Future<RecordMetadata> {
    return producer.send(ProducerRecord<kotlin.String, TreeNode>(topic, null, key, value, headers.toHeaders()))
  }
}